### PR TITLE
[project configuration]: fix tsconfig.json

### PR DESCRIPTION
**Pull request by MarioBanovac**

### CODE REVIEW

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts [hunk 1]
```diff
@@ -1,9 +1,66 @@
...
+  });
 
-  test('hello', () => {
-    expect("between 5 and 15").toContain('5 and 15')
-  })
-})
+  test.each(["password", "12345", "#$%&/"])(
+    "knows that the password: %s is between 5 and 15 characters long",
+    (userInput) => {
+      expect(passwordValidator.hasValidLength(userInput)).toBeTruthy();
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts]: This (hasValidLengrh) is interval API, right? I.e., you don't expect the user of your passwordvalidator to call this. If so, hide it!

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts [hunk 2]
```diff
@@ -1,9 +1,66 @@
...
+    "knows that the password: %s is not between 5 and 15 characters long",
+    (userInput) => {
+      expect(passwordValidator.hasValidLength(userInput)).toBeFalsy();
+    }
+  );
 
+  test.each(["pass1", "1pass", "pass1word$"])(
+    "knows that the password: %s contains a digit",
+    (userInput) => {
+      expect(passwordValidator.containsDigit(userInput)).toBeTruthy();
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts]: Same here, internal API right?

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts [hunk 3]
```diff
@@ -1,9 +1,66 @@
...
+  );
+
+  test.each(["password", "random", "1pass"])(
+    "knows that the password: %s does not contain an uppercase letter",
+    (userInput) => {
+      expect(passwordValidator.containsUppercase(userInput)).toBeFalsy();
+    }
+  );
+
+  test.each([
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts:9]: 👌

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts [hunk 4]
```diff
@@ -1,9 +1,66 @@
...
+    ["pass", { success: false, errors: ["TOO_SHORT", "NO_DIGIT", "NO_UPPERCASE"]}],
+    ["1pas", { success: false, errors: ["TOO_SHORT", "NO_UPPERCASE"]}],
+    ["Password", { success: false, errors: ["NO_DIGIT"]}]
+  ])("knows when the password: %s is not valid", (userInput, expected) => {
+    expect(passwordValidator.isValid(userInput)).toEqual(expected);
+  });
+
+  test.each([
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts:17]: 👌

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    return /\d/.test(userInput);
+  }
+
+  public containsUppercase(userInput: string): boolean {
+    return /[A-Z]/.test(userInput);
+  }
+
+  public isValid(userInput: string) {
+    let returnObject: {
+      success: null | boolean;
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts]: Succes can't be null, right?

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      this.hasValidLength(userInput) &&
+      this.containsDigit(userInput) &&
+      this.containsUppercase(userInput)
+    ) {
+      return {
+        ...returnObject,
+        success: true,
+      };
+    }
+
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:35]: Remove this, in favor of (see next comment)

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      returnObject.errors.push("NO_DIGIT");
+    }
+    
+    if(!this.containsUppercase(userInput)) {
+      returnObject.errors.push("NO_UPPERCASE")
+    }
+    
+    return {
+      ...returnObject,
+      success: false
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:50]:
```suggestion
      success: Boolean(returnObject.length)
```

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+  }
+}
```
Comment (jvhellemondt) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:53]: Isn't the business rule, TOO_LONG missing?

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts (hunk 7 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:50]: I don't understand this comment

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts (hunk 8 above)
Comment (MarioBanovac) [reply ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:53]: Yes

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    
+    if(!this.containsUppercase(userInput)) {
+      returnObject.errors.push("NO_UPPERCASE")
+    }
+    
+    return {
+      ...returnObject,
+      success: false
+    }
```
Comment (enekesabel) [inline ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:51]: @MarioBanovac maybe it's easier to process it in a single diff. @jvhellemondt meant this I think:

```suggestion

    if (!this.hasValidLength(userInput)) {
      returnObject.errors.push("INVALID_LENGTH");
    }

    if (!this.containsDigit(userInput)) {
      returnObject.errors.push("NO_DIGIT");
    }
    
    if(!this.containsUppercase(userInput)) {
      returnObject.errors.push("NO_UPPERCASE")
    }
    
    return {
      ...returnObject,
      success: Boolean(returnObject.length)
    }
```

**Review by enekesabel — APPROVED** [review]
Very neat submission of the TDD exercise all the way. You also did a great job with the upfront design imo.

File: ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts (hunk 7 above)
Comment (jvhellemondt) [reply ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.ts:50]: Sorry, yes, I get that. It’s the next one, mentioned in: https://github.com/MarioBanovac/the-software-essentialist/pull/3#discussion_r1768720891

**Pull request merged**